molgenis-data-security/src/main/java/org/molgenis/data/security/owned/AbstractRowLevelSecurityRepositoryDecorator.java

```diff
@@ -0,0 +1,281 @@
+	public void forEachBatched(Fetch fetch, Consumer<List<E>> consumer, int batchSize)
+	{
+		delegate().forEachBatched(fetch, entities -> consumer.accept(
+				entities.stream().filter(entity -> hasPermissionOnEntity(entity, Operation.READ)).collect(toList())),
+				batchSize);
+	}
+
+	@Override
+	public long count()
+	{
+		return findAllPermitted(Operation.COUNT).count();
+	}
+
+	@Override
+	public long count(Query<E> q)
+	{
+		return findAllPermitted(q, Operation.COUNT).count();
+	}
+
+	@Override
+	public Stream<E> findAll(Query<E> q)
+	{
+		return findAllPermitted(q, Operation.READ);
+	}
+
+	@Override
+	public E findOne(Query<E> q)
+	{
+		E entity = delegate().findOne(q);
+
+		if (entity != null && !hasPermissionOnEntity(entity, Operation.READ))
+		{
+			return null;
+		}
+
+		return entity;
+	}
+
+	@Override
+	public E findOneById(Object id)
+	{
+		E entity = null;
+
+		if (hasPermissionOnEntity(id, Operation.READ))
+		{
+			entity = delegate().findOneById(id);
+		}
+		return entity;
+	}
+
+	@Override
+	public E findOneById(Object id, Fetch fetch)
+	{
+		E entity = null;
+		if (hasPermissionOnEntity(id, Operation.READ))
+		{
+			entity = delegate().findOneById(id, fetch);
+		}
+		return entity;
+	}
+
+	@Override
+	public Stream<E> findAll(Stream<Object> ids)
+	{
+		Stream<E> entities = delegate().findAll(ids);
+		entities = entities.filter(entity -> hasPermissionOnEntity(entity, Operation.READ));
+		return entities;
+	}
+
+	@Override
+	public Stream<E> findAll(Stream<Object> ids, Fetch fetch)
+	{
+		return delegate().findAll(ids, fetch).filter(entity -> hasPermissionOnEntity(entity, Operation.READ));
+	}
+
+	@Override
+	public AggregateResult aggregate(AggregateQuery aggregateQuery)
+	{
+		if (!currentUserIsSuOrSystem())
+		{
+			throw new UnsupportedOperationException();
+		}
+		return delegate().aggregate(aggregateQuery);
+	}
+
+	@Override
+	public void update(E entity)
+	{
+		if (!hasPermissionOnEntity(entity, Operation.UPDATE))
+		{
+			throw new MolgenisDataAccessException(
+					format("No [%s] permission on entity type [%s] with id [%s]", toMessagePermission(Operation.DELETE),
+							entity.getEntityType().getLabel(), entity.getIdValue()));
+		}
+		delegate().update(entity);
+		updateAcl(entity);
+	}
+
+	@Override
+	public void update(Stream<E> entities)
+	{
+		delegate().update(entities.filter((E entity) ->
+		{
+			boolean result = hasPermissionOnEntity(entity.getIdValue(), Operation.UPDATE);
+			if (result)
+			{
+				updateAcl(entity);
+			}
+			return result;
+		}));
+	}
+
+	@Override
+	public void delete(E entity)
+	{
+		if (!hasPermissionOnEntity(entity, Operation.DELETE))
+		{
+			throw new MolgenisDataAccessException(
+					format("No [%s] permission on entity type [%s] with id [%s]", toMessagePermission(Operation.DELETE),
+							entity.getEntityType().getLabel(), entity.getIdValue()));
+		}
+		deleteAcl(entity);
+		delegate().delete(entity);
+	}
+
+	protected abstract String toMessagePermission(Operation operation);
```

Comment:
Does this differ per decorator?

Reply:
Yes since EntityType en Entity have their own set of permissions